src/lib/getClasses.ts

```diff
@@ -14,15 +24,45 @@ const getClasses = (
   renderer: IRenderer,
   componentStyles: IComponentPartStylesInput,
   styleParam: ComponentStyleFunctionParam,
+  extra: any,
 ): IComponentPartClasses => {
   // root, icon, etc.
   const componentParts: string[] = Object.keys(componentStyles)
 
-  return componentParts.reduce((classes, partName) => {
-    classes[partName] = renderer.renderRule(callable(componentStyles[partName]), styleParam)
+  const { className: componentClassName } = extra
+
+  return componentParts.reduce(
+    (classes, partName) => {
+      // handle state section differently
+      // DISCLAIMER: dirty code alert!
+      if (partName.startsWith('on')) {
+        const pseudoStateName = partName.substr(2).toLowerCase()
+
+        const slotStyles = (componentStyles[partName] as any)()
+
+        Object.keys(slotStyles).forEach(slotName => {
+          const slotStyle = callable(slotStyles[slotName])(styleParam)
+          const slotStyleHash = getHashCode(JSON.stringify(slotStyle))
+
+          const slotClassName = `${slotName}-slot-of-${componentClassName}--${slotStyleHash}`
```

Comment:
this closely attributes to your comment about the `Icon` reacting on hover state changes of `Button` that is not its direct parent (`Button` wrapped in `Button` story). Was experimenting with different approaches to address this problem, and to this point preferring the one about introducing notion of `componentId` for parent component, so that slot selector will be like

```css
.${component-class-name}-${componentId} .${slotName}-slot-of-${componentId}--${slotStyleHash}`
```

With that the problem of `Button` -> `Button` -> `Icon` case should be solved - however, obviously, there are several questions that need to be answered before this approach could be proposed as **a** solution:
- what are perf implications (not worrying much about that, frankly)
- !! we should prevent bloat of different `componentId`s - otherwise we will end up with explosion of static styles on the page

Another way to solve the problem is to look on CSS side - specifically, to be smarter in terms of selector to use, so that this problem won't happen. Experimenting with that as well